CTabFolder: Add 'showSelectedImage' property

shubhamWaghmare-sap opened the pull request:
- Added new property "showSelectedImage" to CTabFolder parallel to existing property "showUnselectedImage"

- Image rendering in drawSelected method updated in CTabFolderRenderer with handling the new property "showSelectedImage"

- Added standard padding for text of tabs when tab icon is either not available or is hidden (both for selected and unselected tabs)

- Pre-requisite for [issue](https://github.com/eclipse-platform/eclipse.platform.ui/issues/735) and pull request : https://github.com/eclipse-platform/eclipse.platform.ui/pull/1071

sratz commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java:
```diff
@@ -1406,7 +1411,9 @@ void drawSelected(int itemIndex, GC gc, Rectangle bounds, int state ) {
-			// draw Text
+			// draw Text			
+			//Add some padding before text, if image is not visible
+			if(!parent.showSelectedImage) xDraw += INTERNAL_SPACING;
```
is this not covered by the condition above already where spacing is added ?

HeikoKlare replied:
As far as I understand, the calculation above is for the size of the tab and this is for the placement of the text within it. Still I do not really like that the code is explained via so many comments (even though it is better compared having no explanation). I would prefer to have methods like `addTextPadding(...)` that consistently adds the padding whereever needed (i.e., both in the size calculation as well as in the text placement). That both avoids inconsistencies in the code and makes the code more comprehensible (and the comments obsolete).

shubhamWaghmare-sap replied:
@sratz  As pointed out by @HeikoKlare, to add padding before the text, the width of the tab should also be increased otherwise the text would run out of the tab. Thus this change is to shift (pad) text by a standard width and same width size is added to the tab in the condition above.

@HeikoKlare We are working on extracting this logic into a separate method getTextPadding(...)  for uniform padding calculation when image is not visible.

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolder.java:
```diff
@@ -3694,6 +3707,25 @@ public void setUnselectedImageVisible(boolean visible) {
+ *    <li>ERROR_THREAD_INVALID_ACCESS - if not called from the thread that created the receiver</li>
+ * </ul>
+ *
+ * @since 3.124
```
I think we need a version bump to 3.125 because of API extension here, don't we? 3.124.0 and 3.124.100 won't contain this API.

shubhamWaghmare-sap replied:
Yes. We will bump it in a new change.

HeikoKlare reviewed (commented):
Thanks for this extension.
I guess the naming like `getSelectedImageVisible()` instead of `isSelectedImageVisible()` is because of consistency with the existing API, isn't it? Apart from that, I only have some minor comments.

shubhamWaghmare-sap commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java:
```diff
@@ -318,9 +318,14 @@ protected Point computeSize (int part, int state, GC gc, int wHint, int hHint) {
 						Rectangle bounds = image.getBounds();
-						if ((state & SWT.SELECTED) != 0 || parent.showUnselectedImage) {
-							width += bounds.width;
+						//If image is not visible, tab width should not include image width
```
Sure. We'll re-format it. 
Also, we are working on extracting this logic into a separate method  `getTextPadding(...) ` for uniform padding calculation when image is not visible. This will simplify the code.

shubhamWaghmare-sap commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java:
```diff
@@ -318,9 +318,14 @@ protected Point computeSize (int part, int state, GC gc, int wHint, int hHint) {
+						//If image is not visible, tab width should not include image width
+						//Additionally, tab width is increased by standard padding (to add padding to the text) 
+						if ((state & SWT.SELECTED) != 0){
+							width += parent.showSelectedImage ? bounds.width : INTERNAL_SPACING;						
```
@sratz We are working on extracting this logic into a separate method  `getTextPadding(...) ` for uniform padding calculation when image is not visible. This will simplify the code.

@HeikoKlare The new method `getTextPadding(...) ` will simplify it. We would be modifying the logic to increase the tab width by standard padding, for both selected and unselected tab.

Regarding the padding in line 344, it does not consider whether the image is visible or not. If image is not visible for a tab, there is very small space before the text, which makes it difficult to distinguish between tabs. Thus additional padding is to be added when image of a tab is not visible.

HeikoKlare commented:
@shubhamWaghmare-sap Looks like you have addressed the review comments. Are you still working on this PR or it is already worth to re-review?

shubhamWaghmare-sap commented:
> @shubhamWaghmare-sap Looks like you have addressed the review comments. Are you still working on this PR or it is already worth to re-review?

@HeikoKlare Yes. We have incorporated the review comments in this PR. It is ready to be re-reviewed now. Thanks. 
We do have some open questions like a better name for the method `getTextPadding(...)` added in `CTabFolderRenderer` and the need for text padding for unselected tabs as well, when image is hidden or not available.

Hoping to get answers via discussions on PR. 👍

A short overview of changes:

- Text padding added in both selected and unselected tab when image is hidden or not available
- In case there is no text for the tab, no padding is added.
- Refactoring.

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java:
```diff
@@ -355,6 +356,7 @@ protected Point computeSize (int part, int state, GC gc, int wHint, int hHint) {
 							gc.setFont(gcFont);
 						}
 					}
+					width += getTextPadding(item, state);
```
I would suggest to apply the padding two times to the overall width, so that the text is centered within the tab (as the padding is applied one time to the x offset of the text).

```suggestion
					width += getTextPadding(item, state) * 2;
```

shubhamWaghmare-sap replied:
In that case, code is reverted to adding the the padding twice uniformly to the tab width. 
Also, in case close button is visible for unselected tab, the close button size is still added to tab width. If we remove the close button size, the text gets truncated and the padding would need to be increased.

With the goal to center the text within tab, close button size wouldn't matter much.

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java:
```diff
@@ -355,6 +356,7 @@ protected Point computeSize (int part, int state, GC gc, int wHint, int hHint) {
 					}
+					width += getTextPadding(item, state);
 					if (parent.showClose || item.showClose) {
 						if ((state & SWT.SELECTED) != 0 || parent.showUnselectedClose) {
 							if (width > 0) width += INTERNAL_SPACING;
```
To me, it feels strange if this internal spacing is also applied when no images are shown, as it leads to a large padding right to the text, especially in case the padding has a significantly higher value.

```suggestion
						if (((state & SWT.SELECTED) != 0 && parent.showSelectedImage)
								|| ((state & SWT.SELECTED) == 0 && parent.showUnselectedImage)) {
							if (width > 0) width += INTERNAL_SPACING;
						}
						if ((state & SWT.SELECTED) != 0 || parent.showUnselectedClose) {
```

shubhamWaghmare-sap replied:
The INTERNAL_SPACING here is applied in case the _close_ is visible for tabs. I believe this has to be there irrespective of the current change.
Also, the spacing is applied **only** when no images are shown. The primary objective of applying the padding is to have proper visual separation between tabs when icons/images are not visible.

I believe there is a gap in my understanding.
@HeikoKlare Could you please elaborate a little in what you wanted to convey via this suggested change?

HeikoKlare replied:
Sorry for the confusion. It's not a gap in your understanding, just something must have went wrong when I added the suggestion. The actual suggestion should be different. Let me explain it textually: The code here does two things: first, it adds the `INTERNAL_SPACING` to the width; seconds, it adds the size of the close button. The latter must _always_ be done, that is correct. But the former one is, from my understanding, supposed to ensure that between the text and the close button there is some spacing. However, when icons are disabled there is already a large gap and adding this addition spacing makes the padding to the left and right of the text uneven.

This is a screenshot with additional spacing (current state):
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/755472/b0df66ea-4c92-4e06-b1ce-fa7c53a35a38)

And this without (my proposal):
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/755472/091ed3f7-6343-4767-bcdb-d25455db4171)

It is only a small different, but it makes the text look slightly uncentered (because it actually is uncentered).

One note: my explanation is conflicting with your statement:
> Also, the spacing is applied only when no images are shown.

Maybe I missed something in the code, but my understand is that this code will always be excuted when the tab to be drawn has a close icon.

shubhamWaghmare-sap replied:
@HeikoKlare Thanks for clarifying. The code is now updated accordingly and the tab text is now centered.

`Before`
<img width="450" alt="image" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/57699330/d8eb1519-d567-449a-b9ae-b37d7d6042d5">

`After`
<img width="450" alt="image" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/57699330/1b24b568-f283-4fd7-b0fc-868b2bc911f2">
> Also, the spacing is applied only when no images are shown.

>> Maybe I missed something in the code, but my understand is that this code will always be excuted when the tab to be drawn has a close icon.

You are right. The code here gets executed when tab has a close icon. I misunderstood it before.

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java:
```diff
@@ -370,6 +372,27 @@ protected Point computeSize (int part, int state, GC gc, int wHint, int hHint) {
+		if (text != null && parent.getMinimumCharacters() != 0) {
+			if (image == null || image.isDisposed() || ((state & SWT.SELECTED) != 0 && !parent.showSelectedImage)
+					|| ((state & SWT.SELECTED) == 0 && !parent.showUnselectedImage))
+				return INTERNAL_SPACING;
```
See general comment: I propose to define the padding when not showing icons separately.
```suggestion
				return TABS_WITHOUT_ICONS_PADDING;
```

HeikoKlare commented on tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/Test_org_eclipse_swt_custom_CTabFolder.java:
```diff
@@ -468,6 +468,20 @@ public void test_iconWrappedOnNextLine() {
+public void test_selectedImageVisible() {
+	createTabFolder(null);
+
+	ctabFolder.setSelectedImageVisible(true);
+	assertTrue( ctabFolder.getSelectedImageVisible() );
+
+	ctabFolder.setSelectedImageVisible(false);
+	assertFalse( ctabFolder.getSelectedImageVisible());
+
```
Please correct formatting here. There are strange whitespaces that will be corrected with auto-formatting.

HeikoKlare reviewed (commented):
Thanks for the improvements to this PR and sorry for the late reply.

In general, I am fine with the changes. I only have two concerns with respect the padding:
1. I would propose to separate the padding for the case that icons are not shown from the existing `INTERAL_SPACING` (in a variable such as `TABS_WITHOUT_ICONS_PADDING`) as they apply to different behaviors and should thus not influence each other.
2. I would even be in favor of having a bigger spacing when no icons are shown. This seems to require 1. to work properly (see my detailed comments).

To me, a value like 14 for such a `TABS_WITHOUT_ICONS_PADDING` feels good :
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/755472/8a0a882c-c3f0-4b9f-8ace-ec4ca970cc2c)
But that may be a matter of personal preference, it might also be smaller but at least should be somehow independent from the existing spacing.

HeikoKlare commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java:
```diff
@@ -355,9 +357,15 @@ protected Point computeSize (int part, int state, GC gc, int wHint, int hHint) {
+							if (((state & SWT.SELECTED) != 0 && parent.showSelectedImage)
+									|| ((state & SWT.SELECTED) == 0 && parent.showUnselectedImage)) {
+								if (width > 0) width += INTERNAL_SPACING;
+							}
```
This is rather nitpicky, but it feels to me like there needs to be less spacing here when no images are shown:
```suggestion
							} else {
								if (width > 0) width -= INTERNAL_SPACING;
							}
```
Actually, it is not about `INTERNAL_SPACING` to be subtracted but about the margin of the close button (which is equal to `INTERNAL_SPACING = 4`). This feels more natural to me, as I would expect the distance between left end of the tab and the beginning of the text should be equal to the distance between the end of the text and the beginning of the icon (and not until the beginning of the icon's margin).

shubhamWaghmare-sap replied:
Incorporated the suggestion and following is how it looks like after the change:
<img width="500" alt="image" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/57699330/9f7e3c82-a3ab-4553-ac15-5d4cf5cf0bed">

HeikoKlare reviewed (commented):
The changes look good to me now. I have one nitpicky finding with respect to the text padding left.
@shubhamWaghmare-sap please also squash all your commits into one in preparation of merging this PR.
@sratz you have an open change request. Have your concerns be addressed or do you want to re-review?

Finally, I have one comment left but only for the sake of documentation, as I do not have a solution for that, so it should not affect this PR:
To me the behavior when showing the close button for unselected tabs does not yet feel right for me. The padding and text placement calculation is equal to when the close button is shown (for the selected tab). This is fine for the case when hovering over an unselected tab, but does not feel right for the other tabs (consider the "Tip of the Day" tab in the following screenshot). But since only one of these cases can be optimized without having tab headers jumping around, I do not see a better solution for this.
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/755472/6ef7c708-b3da-422c-930a-47ea617f3ec1)

HeikoKlare reviewed (approved):
Thanks for this great contribution. Content-wise, the changes are ready to be merged from my side.

@shubhamWaghmare-sap Can you please squash the commits in this PR into one?

@sratz Can we remove your change request or is there anything left to do?

sratz commented:
Looks good.

@shubhamWaghmare-sap: Can you rebase on top of master?

shubhamWaghmare-sap commented:
@sratz @HeikoKlare 
Thanks for all the comments and suggestions and approving the changes.
I have squashed all commits and re-based on top of master.

HeikoKlare commented:
@shubhamWaghmare-sap Thank for this contribution! Looking forward to having the Platform UI PR merged and the feature available in Eclipse.

fedejeanne commented on bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/CTabFolder.java:
```diff
@@ -1366,6 +1367,18 @@ public boolean getUnselectedImageVisible() {
+ *
+ * @return <code>true</code> if an image appears in selected tabs
+ *
+ * @since 3.125
```
Shouldn't the Manifest and the Pom also be updated in this case?

HeikoKlare replied:
The required version bump originally was part of this PR: https://github.com/eclipse-platform/eclipse.platform.swt/commit/cf4691d7d546f70b038a0c9b9bf74fe5603c219f#diff-6485b884863fd4f54a20976acef942fe0dd257eb4e72ca25533fd84b7b6fe2e6

But it obviously got lost. Interesting that the API tools baseline comparison did not complain.

I provided #944 to have the faulty version fixed soon.

HeikoKlare replied:
Fixed

iloveeclipse replied:
Here is related snippet from build.
https://ci.eclipse.org/releng/job/eclipse.platform.swt/job/PR-785/12/consoleText
@laeubi : I don't see API tools reporting anything except there is no baseline??

```
[2023-12-14T11:04:05.054Z] [INFO] --- tycho-source:4.0.5-SNAPSHOT:plugin-source (plugin-source) @ org.eclipse.swt ---
[2023-12-14T11:04:05.054Z] [INFO] 
[2023-12-14T11:04:05.054Z] [INFO] --- tycho-packaging:4.0.5-SNAPSHOT:package-plugin (default-package-plugin) @ org.eclipse.swt ---
[2023-12-14T11:04:05.311Z] [INFO] Building jar: /home/jenkins/agent/workspace/eclipse.platform.swt_PR-785/eclipse.platform.swt/bundles/org.eclipse.swt/target/org.eclipse.swt-3.124.300-SNAPSHOT.jar
[2023-12-14T11:04:05.577Z] [INFO] 
[2023-12-14T11:04:05.577Z] [INFO] --- tycho-p2:4.0.5-SNAPSHOT:p2-metadata-default (default-p2-metadata-default) @ org.eclipse.swt ---
[2023-12-14T11:04:05.578Z] [INFO] No baseline version MavenProject: org.eclipse.swt:org.eclipse.swt:3.124.300-SNAPSHOT @ /home/jenkins/agent/workspace/eclipse.platform.swt_PR-785/eclipse.platform.swt/bundles/org.eclipse.swt/pom.xml
[2023-12-14T11:04:05.578Z] [INFO] 
[2023-12-14T11:04:05.578Z] [INFO] --- tycho-p2-extras:4.0.5-SNAPSHOT:compare-version-with-baselines (compare-attached-artifacts-with-release) @ org.eclipse.swt ---
[2023-12-14T11:04:05.578Z] [INFO] Adding repository https://download.eclipse.org/eclipse/updates/4.30/R-4.30-202312010110
[2023-12-14T11:04:05.835Z] [INFO] 
[2023-12-14T11:04:05.835Z] [INFO] --------< org.eclipse.swt:org.eclipse.swt.fragments.localbuild >--------
```

laeubi replied:
>  don't see API tools reporting anything except there is no baseline??

SWT projects have no comparable baseline because of the way they are build. This might change when LSF is enabled (or submodules) and we have one repository as then it would be possible to transform them into regular (tycho) projects.

HannesWell replied:
Some work and expermients needs to be done, but I have the intention to do that during the Christmas new-years holidays.

HeikoKlare replied:
Great to hear that you will work on that, @HannesWell, but don't forget to also use holidays as holidays 🙂

merks commented:
Is it an unintended side effect that the tabs are so much wider now?

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/c68b2ee6-19d9-45fe-81f7-70e31cea996f)

It doesn't look very nice...

HeikoKlare commented:
> Is it an unintended side effect that the tabs are so much wider now?

Yes. it is.

> It doesn't look very nice...

True.

The larger padding is intended to applied to tab folders that usually contain images but that may be disabled (such as tab folder of Eclipse views: eclipse-platform/eclipse.platform.ui#1071). But it currently applies to all tabs for which showing images is disabled _or_ that have no image assigned:

https://github.com/eclipse-platform/eclipse.platform.swt/blob/96692a51e54863a4670d88a78f0ab12043a07fa3/bundles/org.eclipse.swt/Eclipse%20SWT%20Custom%20Widgets/common/org/eclipse/swt/custom/CTabFolderRenderer.java#L396-L398

The latter is true for all existing tab folders without images, such as the one in the screenshot. So the functionality added with this PR has to be improved, either by only applying the padding when images are explicitly disabled or by adding a propery for a tab folder that defines whether the new behavior (no images but larger padding) shall be enabled.

sratz commented:
> either by only applying the padding when images are explicitly disabled or by adding a propery for a tab folder that defines whether the new behavior (no images but larger padding) shall be enabled.

Alternatively: Would it make sense to expose

> TABS_WITHOUT_ICONS_PADDING

as a method rather than a constant (with backwards-compatible default value), such that the subclass in platform.ui can control (and potentially fine-tune via preferences) this beheavior?

HeikoKlare commented:
> Alternatively: Would it make sense to expose
> 
> > TABS_WITHOUT_ICONS_PADDING
> 
> as a method rather than a constant (with backwards-compatible default value), such that the subclass in platform.ui can control (and potentially fine-tune via preferences) this behavior?

Making the value configurable (potentially also via preferences) sound like a good idea to me. I was not aware that platform.ui has a specialization of this renderer, so providing the padding value via method and overwriting it in platform.ui would be possible. It could also be made "dynamically" configurable via some setter logic. Then you would not need to make a subclass for overwriting the value but can just do that by instantiating the renderer.
